### PR TITLE
feat: add `Series|Expr.rolling_sum` method

### LINKED ISSUE #1367: api: "unstable" features
I'm slightly hesitant about how to merge #1298 and #1290 - rolling_mean is marked as unstable in Polars. ewm_mean isn't, but follows the api kind of API style

I do think we should merge them, I just think we need a way to signal that we won't make the same kinds of promises for them as we do for the rest of the API in `stable.v1`

So, I'm tempted to say:
- let's wait for some decision / further discussion on https://github.com/plotly/plotly.py/issues/4834
- we can discuss the Narwhals unstable API

I'm sorry for the delay here in merging these PRs

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Related issue #1254
- Closes #1367 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below

So I wanted to start with `sum`, assuming it would have been simpler for arrow to implement in a way which is not naive.

Running some benchmarks, performances of this with 1M rows is in the same ballpark of pandas, while the naive way in #1290 is orders of magnitude slower. I would consider this the way to go forward.

### CODE REVIEW

File: narwhals/exceptions.py [hunk 1]
```diff
@@ -55,3 +55,7 @@ def from_invalid_type(cls, invalid_type: type) -> InvalidIntoExprError:
             "named `0`."
         )
         return InvalidIntoExprError(message)
+
+
+class NarwhalsUnstableWarning(UserWarning):
+    """Warning issued when a method or function is considered unstable in the stable api."""
```
Comment (FBruzzesi) [inline narwhals/exceptions.py:61]: Should this be exposed?

Comment (MarcoGorelli) [reply narwhals/exceptions.py:61]: i think so yes

File: narwhals/stable/v1/__init__.py [hunk 2]
```diff
@@ -492,11 +493,191 @@ def value_counts(
...
+            ]
+        """
+        from narwhals.exceptions import NarwhalsUnstableWarning
+        from narwhals.utils import find_stacklevel
+
+        msg = (
+            "`Series.rolling_sum` is being called from the stable API although considered "
+            "an unstable feature."
+        )
+        warn(message=msg, category=NarwhalsUnstableWarning, stacklevel=find_stacklevel())
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply narwhals/stable/v1/__init__.py:578]: yup, nice!

**Review by MarcoGorelli — COMMENTED** [review]
wow, this is clever! do you have an idea for how to do the mean / min / max cases?

**Comment by MarcoGorelli** [comment]
Love the creativity here

I tried running this hypothesis test
```python
from hypothesis import given
import hypothesis.strategies as st
import pyarrow as pa
import pandas as pd

@given(
    center = st.booleans(),
    values = st.lists(st.floats(-10, 10), min_size=3, max_size=10),
)
@pytest.mark.filterwarnings('ignore:.*:narwhals.exceptions.NarwhalsUnstableWarning')
def test_rolling_sum_hypothesis(center: bool, values: list[float]) -> None:
    s = pd.Series(values)
    n_missing = random.randint(0, len(s)-1)
    window_size = random.randint(1, len(s))
    min_periods = random.randint(0, window_size)
    mask = random.sample(range(len(s)), n_missing)
    s[mask] = None
    df = pd.DataFrame({'a': s})
    expected = s.rolling(window=window_size, center=center, min_periods=min_periods).sum().to_frame('a')
    result = nw.from_native(pa.Table.from_pandas(df)).select(nw.col('a').rolling_sum(window_size, center=center, min_periods=min_periods))
    expected_dict = nw.from_native(expected, eager_only=True).to_dict(as_series=False)
    assert_equal_data(result, expected_dict)
```
and it's picking up some small inconsistencies:

```
In [19]: s
Out[19]: 
0    0.0
1    NaN
2    0.0
dtype: float64

In [20]: s.rolling(min_periods=0, center=False, window=2).sum()
Out[20]: 
0    0.0
1    0.0
2    0.0
dtype: float64

In [21]: nw.from_native(pa.chunked_array([s]), series_only=True).rolling_sum(min_periods=0, center=False, window_size=2).to_native()
Out[21]: 
<pyarrow.lib.ChunkedArray object at 0x7f7cd5793ee0>
[
  [
    null,
    null,
    null
  ]
]

In [22]: nw.from_native(pl.from_pandas(s), series_only=True).rolling_sum(min_periods=0, center=False, window_size=2).to_native()
Out[22]: 
shape: (3,)
Series: '' [f64]
[
        0.0
        0.0
        0.0
]
```

**Comment by FBruzzesi** [comment]
> Love the creativity here
> 
> I tried running this hypothesis test
> 
> ```python
> from hypothesis import given
> import hypothesis.strategies as st
> import pyarrow as pa
> import pandas as pd
> 
> @given(
>     center = st.booleans(),
>     values = st.lists(st.floats(-10, 10), min_size=3, max_size=10),
> )
> @pytest.mark.filterwarnings('ignore:.*:narwhals.exceptions.NarwhalsUnstableWarning')
> def test_rolling_sum_hypothesis(center: bool, values: list[float]) -> None:
>     s = pd.Series(values)
>     n_missing = random.randint(0, len(s)-1)
>     window_size = random.randint(1, len(s))
>     min_periods = random.randint(0, window_size)
>     mask = random.sample(range(len(s)), n_missing)
>     s[mask] = None
>     df = pd.DataFrame({'a': s})
>     expected = s.rolling(window=window_size, center=center, min_periods=min_periods).sum().to_frame('a')
>     result = nw.from_native(pa.Table.from_pandas(df)).select(nw.col('a').rolling_sum(window_size, center=center, min_periods=min_periods))
>     expected_dict = nw.from_native(expected, eager_only=True).to_dict(as_series=False)
>     assert_equal_data(result, expected_dict)
> ```
> 
> and it's picking up some small inconsistencies:
> 
> ```
> In [19]: s
> Out[19]: 
> 0    0.0
> 1    NaN
> 2    0.0
> dtype: float64
> 
> In [20]: s.rolling(min_periods=0, center=False, window=2).sum()
> Out[20]: 
> 0    0.0
> 1    0.0
> 2    0.0
> dtype: float64
> 
> In [21]: nw.from_native(pa.chunked_array([s]), series_only=True).rolling_sum(min_periods=0, center=False, window_size=2).to_native()
> Out[21]: 
> <pyarrow.lib.ChunkedArray object at 0x7f7cd5793ee0>
> [
>   [
>     null,
>     null,
>     null
>   ]
> ]
> 
> In [22]: nw.from_native(pl.from_pandas(s), series_only=True).rolling_sum(min_periods=0, center=False, window_size=2).to_native()
> Out[22]: 
> shape: (3,)
> Series: '' [f64]
> [
>         0.0
>         0.0
>         0.0
> ]
> ```

Well ok `min_periods = min_periods or window_size` evaluates to 2, because of `0 or 2`. Let me adjust

File: tests/expr_and_series/rolling_sum_test.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    },
+    "x5": {
+        "kwargs": {"window_size": 4, "min_periods": 1, "center": True},
+        "expected": [1.0, 3.0, 3.0, 7.0, 12.0, 21.0, 21.0],
+    },
+    "x6": {
+        "kwargs": {"window_size": 0},
+        "expected": [float("nan"), 1.0, 3.0, 3.0, 7.0, 13.0, 24.0],
+    },
+    # There are still some edge cases to take care of with nulls and min_periods=0:
```
Comment (FBruzzesi) [inline tests/expr_and_series/rolling_sum_test.py:37]: Example:

```py
In [1]: import pandas as pd

In [2]: import polars as pl

In [3]: data = [float("nan"), 1, 2]

In [4]: pl.Series(data).rolling_sum(2, min_periods=0)
Out[4]: 
shape: (3,)
Series: '' [f64]
[
        NaN
        NaN
        3.0
]

In [5]: pd.Series(data).rolling(2, min_periods=0).sum()
Out[5]: 
0    0.0
1    1.0
2    3.0
dtype: float64
```

Comment (FBruzzesi) [reply tests/expr_and_series/rolling_sum_test.py:37]: Honestly, this seems buggy in both?

- polars: why the second value is NaN if `min_periods=0`?
- pandas: why replace the first value with a 0?

**Comment by FBruzzesi** [comment]
I would tend towards restricting the API to:

- `window_size` strictly positive
- `min_periods` either None or strictly positive

and raise in other cases.

File: tests/expr_and_series/rolling_sum_test.py (hunk 3 above)
Comment (MarcoGorelli) [reply tests/expr_and_series/rolling_sum_test.py:37]: in Polars `min_periods` refers to missing data, not to `'nan'`

the mean of `'nan'` and 1 is '`nan'`

**Comment by MarcoGorelli** [comment]
thanks - shall we also include the hypothesis test?

File: tests/expr_and_series/rolling_sum_test.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    with context:
+        df["a"].rolling_sum(window_size=window_size, min_periods=min_periods)
+
+
+@given(  # type: ignore[misc]
+    center=st.booleans(),
+    values=st.lists(st.floats(-10, 10), min_size=3, max_size=10),
+)
+@pytest.mark.filterwarnings("ignore:.*:narwhals.exceptions.NarwhalsUnstableWarning")
+def test_rolling_sum_hypothesis(center: bool, values: list[float]) -> None:  # noqa: FBT001
```
Comment (MarcoGorelli) [inline tests/expr_and_series/rolling_sum_test.py:201]: shall we skip this for pandas < 1?

**Review by MarcoGorelli — APPROVED** [review]
very impressive @FBruzzesi , well done
